custom-recipes/object-detection-query-sampler/recipe.py

```diff
@@ -0,0 +1,487 @@
+            filtered_scores = keras.backend.gather(scores, indices)[:, 0]
+
+            # perform NMS
+            nms_indices = backend.non_max_suppression(filtered_boxes, filtered_scores, max_output_size=max_detections, iou_threshold=nms_threshold)
+
+            # filter indices based on NMS
+            indices = keras.backend.gather(indices, nms_indices)
+
+        # add indices to list of all indices
+        labels = backend.gather_nd(labels, indices)
+        indices = keras.backend.stack([indices[:, 0], labels], axis=1)
+
+        return indices
+
+    if class_specific_filter:
+        all_indices = []
+        # perform per class filtering
+        for c in range(int(classification.shape[1])):
+            scores = classification[:, c]
+            labels = c * backend.ones((keras.backend.shape(scores)[0],), dtype='int64')
+            all_indices.append(_filter_detections(scores, labels))
+
+        # concatenate indices to single tensor
+        indices = keras.backend.concatenate(all_indices, axis=0)
+    else:
+        scores  = keras.backend.max(classification, axis    = 1)
+        labels  = keras.backend.argmax(classification, axis = 1)
+        indices = _filter_detections(scores, labels)
+
+    # select top k
+    cscores = keras.backend.gather(classification, indices[:, 0])
+    scores              = backend.gather_nd(classification, indices)
+    labels              = indices[:, 1]
+    scores, top_indices = backend.top_k(scores, k=keras.backend.minimum(max_detections, keras.backend.shape(scores)[0]))
+
+    # filter input using the final set of indices
+    indices             = keras.backend.gather(indices[:, 0], top_indices)
+    boxes               = keras.backend.gather(boxes, indices)
+    labels              = keras.backend.gather(labels, top_indices)
+    other_              = [keras.backend.gather(o, indices) for o in other]
+
+    # Take predictions of all classes
+    scores = keras.backend.gather(cscores, top_indices)
+    
+    # zero pad the outputs
+    pad_size = keras.backend.maximum(0, max_detections - keras.backend.shape(scores)[0])
+    boxes    = backend.pad(boxes, [[0, pad_size], [0, 0]], constant_values=-1)
+    scores   = backend.pad(scores, [[0, pad_size], [0, 0]], constant_values=-1)
+    labels   = backend.pad(labels, [[0, pad_size]], constant_values=-1)
+    labels   = keras.backend.cast(labels, 'int32')
+    other_   = [backend.pad(o, [[0, pad_size]] + [[0, 0] for _ in range(1, len(o.shape))], constant_values=-1) for o in other_]
+
+    # set shapes, since we know what they are
+    boxes.set_shape([max_detections, 4])
+    scores.set_shape([max_detections, int(classification.shape[1])])
+    labels.set_shape([max_detections])
+    for o, s in zip(other_, [list(keras.backend.int_shape(o)) for o in other]):
+        o.set_shape([max_detections] + s[1:])
+
+    return [boxes, scores, labels] + other_
+
+
+class MyFilterDetections(FilterDetections):
```

Comment:
I'm missing how is this different from the original [FilterDetection](https://github.com/fizyr/keras-retinanet/blob/f4017fdfb942ef594bf3f81d846f6e6100fa9e5d/keras_retinanet/layers/filter_detections.py#L115) ?